### DISCUSSION TITLE
Crash - How can the WinUI3 application be opened using a protocol?

### DISCUSSION THREAD

**Discussion by stefanvd** [microsoft/WindowsAppSDK#3795] [General]
Hi there,

I have 2 WinUI3 applications in a package project, both with protocol schemes such as "settings" and "main." However, when I try to open the protocol using the Run window with "settings://," it crashes.
```
System.Runtime.InteropServices.COMException: 'The application
called an interface that was marshalled for a different thread.
'
```

I'm using 2 "blank app packaged" WinUI3 projects and 1 "packaged".

Do you have any ideas on how to solve this issue? Or perhaps a working sample that I could refer to?

<img width="1556" alt="Screenshot 2023-08-17 at 10 35 41" src="https://github.com/microsoft/WindowsAppSDK/assets/20127018/fe774364-eb69-4134-b967-90fc0b92efb3">

<img width="1559" alt="Screenshot 2023-08-17 at 10 35 54" src="https://github.com/microsoft/WindowsAppSDK/assets/20127018/494f9c45-a4b8-4969-bf64-8d7234e4f343">

**Comment by DarranRowe** [comment]
I know that the tools don't work that well with multiple applications per package. But I know for sure that the "Entry point" setting isn't what you think it is. According to the uap:Extension documentation, it is a WinRT type, so it would be like the entry point for the application, not the ID for the applicaion.
If you look at the raw XML in the package manifest:
```
  <Applications>
    <Application Id="App" Executable="$targetnametoken$.exe" EntryPoint="$targetentrypoint$">
      <uap:VisualElements DisplayName="ActTest (Package)" Description="ActTest (Package)" BackgroundColor="transparent" Square150x150Logo="Images\Square150x150Logo.png" Square44x44Logo="Images\Square44x44Logo.png">
        <uap:DefaultTile Wide310x150Logo="Images\Wide310x150Logo.png" />
        <uap:SplashScreen Image="Images\SplashScreen.png" />
      </uap:VisualElements>
      <Extensions>
        <uap:Extension Category="windows.protocol" >
          <uap:Protocol Name="settings">
            <uap:DisplayName>Settings protocol</uap:DisplayName>
          </uap:Protocol >
        </uap:Extension>
      </Extensions>
    </Application>
  </Applications>
```
The extension is grouped with an application. I would honestly suggest you look at the XML to see what is going on, and make sure that the protocols are grouped with the application it is meant to activate.

Finally, there aren't any samples available for multiple applications in a single package. But the [Windows App SDK samples](https://github.com/microsoft/WindowsAppSDK-Samples/tree/main/Samples/AppLifecycle/Activation) have a single application in the package, for the packaged versions. None of these specify an executable or an entry point. This means that it gets the activation information from the application that it is grouped with.

**Reply by stefanvd** [reply]
Hi there,

Thank you for your response. However, after removing the "entry point," nothing seems to have changed. An error message now indicates that there is no application available for the "settings://" protocol. 
<img width="536" alt="Screenshot 2023-08-17 at 21 14 08" src="https://github.com/microsoft/WindowsAppSDK/assets/20127018/64b10f4c-89e4-4356-bfd0-9b7badd4e740">

Do I need to add something else?

This is my code, in the "settings" app (Package.appxmanifest):

```
  <Applications>
    <Application Id="App"
      Executable="$targetnametoken$.exe"
      EntryPoint="$targetentrypoint$">
      <uap:VisualElements
        DisplayName="Settings"
        Description="Settings"
        BackgroundColor="transparent"
        Square150x150Logo="Assets\Square150x150Logo.png"
        Square44x44Logo="Assets\Square44x44Logo.png">
        <uap:DefaultTile Wide310x150Logo="Assets\Wide310x150Logo.png"  Square71x71Logo="Assets\SmallTile.png" Square310x310Logo="Assets\LargeTile.png"/>
        <uap:SplashScreen Image="Assets\SplashScreen.png" />
      </uap:VisualElements>
      <Extensions>
        <uap:Extension Category="windows.protocol">
          <uap:Protocol Name="settings"/>
        </uap:Extension>
      </Extensions>
    </Application>
  </Applications>
```

And in the packaging (Package.appxmanifest):
```
  <Applications>
    <Application Id="App"
      Executable="$targetnametoken$.exe"
      EntryPoint="$targetentrypoint$">
      <uap:VisualElements
        DisplayName="Package"
        Description="Package"
        BackgroundColor="transparent"
        Square150x150Logo="Images\Square150x150Logo.png"
        Square44x44Logo="Images\Square44x44Logo.png">
        <uap:DefaultTile Wide310x150Logo="Images\Wide310x150Logo.png" />
        <uap:SplashScreen Image="Images\SplashScreen.png" />
      </uap:VisualElements>
    </Application>
  </Applications>
```

**Comment by DarranRowe** [comment]
No.

https://github.com/microsoft/WindowsAppSDK/assets/52577874/9aeac4ad-0311-4599-b87e-9debd00ecd7f

The manifest is what you see there. The code for this is:

```
    MainWindow::MainWindow()
    {
        InitializeComponent();

        auto instance = Microsoft::Windows::AppLifecycle::AppInstance::GetCurrent();
        auto activated_event_args = instance.GetActivatedEventArgs();
        if (activated_event_args != nullptr)
        {
            auto activated_event_args_kind = activated_event_args.Kind();
            if (activated_event_args_kind == Microsoft::Windows::AppLifecycle::ExtendedActivationKind::Protocol)
            {
                auto protocol_activated_event_args = activated_event_args.Data().as<Windows::ApplicationModel::Activation::IProtocolActivatedEventArgs>();
                _ASSERTE(protocol_activated_event_args != nullptr);
                myTextBlock().Text(protocol_activated_event_args.Uri().AbsoluteCanonicalUri());
            }
            else
            {
                _ASSERTE(activated_event_args_kind == Microsoft::Windows::AppLifecycle::ExtendedActivationKind::Launch);
                myTextBlock().Text(L"Not Protocol Activated.");
            }
        }
    }
```
Sure, it is C++, but you should get the idea. The Xaml window only has a text block control named myTextBlock.

--Edit--
Looking again, I did set the display name. I wonder if that is needed.

**Comment by stefanvd** [comment]
Thank you once again for your response. However, I believe you are only using 1 app and 1 package. The issue actually involves 3 projects ("main," "settings," and "package"). By default in Visual Studio, I run the "package" project, and by default, it runs the "main" project (=the target entry point). Nevertheless, when I type or click a button in my "main" app ("settings://hello"), it should open the other "settings" window. But still see the same crash issue.
You can find my code on GitHub:
https://github.com/stefanvd/winui3protocol

**Reply by DarranRowe** [reply]
Well, there is definitely weird things going on under the surface. While doing some experiments, I have gotten things to crash in some really interesting ways. But that isn't important here.
As a simple suggestion, now that I know precisely what you are doing, have you tried setting the entry point to windows.FullTrustApplication?

https://github.com/microsoft/WindowsAppSDK/assets/52577874/fbd24177-3673-46bf-ae25-450faff3de1b

Also, just to be sure, have you made sure that you are pointing to the right location in the package. 

![Screenshot 2023-08-19 010800](https://github.com/microsoft/WindowsAppSDK/assets/52577874/bd08312e-18f1-4f1d-8cf2-190cb45ecbc8)

As you can see, the application that handles settings isn't in a subdirectory.

**Reply by DarranRowe** [reply]
I managed to look into things a little more, and honestly, I would suggest you create a bug report for this on the [WinUI repository](https://github.com/microsoft/microsoft-ui-xaml).
This is failing with the error:
```
Exception thrown: 'System.Runtime.InteropServices.COMException' in WinRT.Runtime.dll
WinRT information: The Application Object must initially be accessed from the multi-thread apartment.
An unhandled exception of type 'System.Runtime.InteropServices.COMException' occurred in WinRT.Runtime.dll
WinRT information: The Application Object must initially be accessed from the multi-thread apartment.
The application called an interface that was marshalled for a different thread.
```
This is part of the "weird things going on under the surface" that I mentioned. Basically, if you set a package up to have two applications defined, and the second application is a WinUI 3 application, then it will consistently fail to start with this error. Setting the main thread to MTA doesn't help since the process will just fail to start and produce a 0x80040904 HRESULT. I have no idea how the test project that I made avoided this either.
The reason why I managed to get this was that I overrode the Main function and disabled the generated one using DISABLE_XAML_GENERATED_MAIN. I then put the main thread to sleep for around 10 seconds. This gave me enough time to attach the debugger.

**Reply by stefanvd** [reply]
I reported this on a new issue report, which can be found at https://github.com/microsoft/microsoft-ui-xaml/issues/8763
